state/protocol/events/gadgets/views.go

```diff
@@ -48,22 +71,28 @@ func (v *Views) BlockFinalized(block *flow.Header) {
 	v.Lock()
 	defer v.Unlock()
 
-	v.lastFinalizedView = block.View
+	blockView := block.View
 
-	// TODO index queued callbacks by ordered view to avoid this linear iteration on each block
-	finalizedViews := []int{}
-	for view := range v.callbacks {
-		if view <= block.View {
-			finalizedViews = append(finalizedViews, int(view))
+	// the index (inclusive) of the lowest view which should be kept
+	cutoff := 0
+	for i, view := range v.orderedViews {
+		if view > blockView {
+			break
 		}
+		for _, callback := range v.callbacks[view] {
+			callback(block)
+		}
+		delete(v.callbacks, view)
+		cutoff = i + 1
 	}
 
-	sort.Ints(finalizedViews)
-
-	for _, view := range finalizedViews {
-		for _, callback := range v.callbacks[uint64(view)] {
-			callback(block)
+	// we have no other queued view callbacks
+	if cutoff >= len(v.orderedViews) {
+		if len(v.orderedViews) > 0 {
+			v.orderedViews = []uint64{}
+			return
 		}
```

Comment:
```suggestion
		if len(v.orderedViews) > 0 {
			v.orderedViews = []uint64{}
		}
		return
```

Should we do this instead?
otherwise, when `len(v.orderedViews) == 0 `, executing L97 will crash due to out of indices

Reply:
I think they are equivalent. L97 would be safe (though unnecessary) because `cutoff` must be `0` if `len(v.orderedViews) == 0`. `someSlice[0:]` is safe when `someSlice` is empty or `nil`. 

Committing the suggestion though, because it is clearer 👍